### PR TITLE
Run mlm pad to multiple for fp16

### LINKED ISSUE #10627: considering `pad_to_multiple_of` for run_mlm.py
## Environment info

- `transformers` version: 4.3.3
- Platform: linux
- Python version: 3.7
- PyTorch version (GPU?): 1.8
- Tensorflow version (GPU?): - 
- Using GPU in script?: - 
- Using distributed or parallel set-up in script?: - 

### Who can help
@sgugger, @patil-suraj

## Information
I have seen in huggingface codes such as run_seq2seq.py that for the case of fp16, they pad to multiple of 8 like below, perhpas for efficiency purpose:

```
 data_collator = DataCollatorForSeq2Seq(
        tokenizer,
        label_pad_token_id=label_pad_token_id,
        pad_to_multiple_of=8 if training_args.fp16 else None,
       )
```

For run_mlm.py codes, the padding condition in case of fp16 is not considered, could it be a bug and possibly performance could get better in run_mlm.py if this was set? kindly appreciate  having a look.

thanks

### PR DESCRIPTION
**Pull request by ak314**
# What does this PR do?
This PR uses padding to a multiple of 8 in the run_mlm.py language modeling example, when fp16 is used. Since the DataCollatorForLanguageModeling did not initially accept the pad_to_multiple_of option, that functionality was added.
Fixes #10627 

## Before submitting
- [X] Did you write any new necessary tests?

## Who can review?
@sgugger

### CODE REVIEW

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -201,7 +201,7 @@ def _collate_batch(examples, tokenizer):
     # Check if padding is necessary.
     length_of_first = examples[0].size(0)
     are_tensors_same_length = all(x.size(0) == length_of_first for x in examples)
-    if are_tensors_same_length:
+    if are_tensors_same_length and (pad_to_multiple_of is None or length_of_first % pad_to_multiple_of == 0):
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:204]: Nice catch!

**Review by sgugger — APPROVED** [review]
Thanks for opening a clean PR!

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -213,6 +213,8 @@ def _collate_batch(examples, tokenizer):
 
     # Creating the full tensor and filling it with our data.
     max_length = max(x.size(0) for x in examples)
+    if pad_to_multiple_of is not None and (max_length % pad_to_multiple_of != 0):
+        max_length = ((max_length // pad_to_multiple_of) + 1) * pad_to_multiple_of
```
Comment (ak314) [inline src/transformers/data/data_collator.py:217]: @sgugger I was considering: typically max model token lengths are already multiples of 8 so it should not be a problem in practice, but in principle the pad to 8x in this method could make the `max_length` bigger then the max model length if that has some strange value. Do you think it's worth adding a check/truncation for that?

Comment (sgugger) [reply src/transformers/data/data_collator.py:217]: I don't think we would ever get a problem with it for the reason you mention (all models have max length that are multiple of 8) and if it did happen it's easy for the user to set a lower max length in the scripts.

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @ak314!

Very nice tests :)
